### PR TITLE
Socket branch

### PR DESCRIPTION
**Pull request by DrJessop**
This pull request includes sockets to update progress information on a particular job, embedded video in web browser, updated UI, potential synchronicity issues are fixed. 
image
image
image
image

### CODE REVIEW

**Comment by abrichr** [comment]
@DrJessop This looks great so far! Can you please add a spinner to the video to indicate that it's still processing?

e.g. https://spin.js.org/

File: .env [hunk 1]
```diff
@@ -1,3 +1,3 @@
 PORT=5000
 POSTGRES_PASSWORD=password
-REDIS_URL=redis://redis:6379
\ No newline at end of file
+REDIS_URL=redis://redis:6379
```
Comment (abrichr) [inline .env:3]: Why is it necessary to add this here now if RQ was able to connect without it before? Can we find out how RQ is getting the connection string and do it the same way? (My concern is having  more than one place where we have to update connection information.)

File: app.py [hunk 2]
```diff
@@ -2,27 +2,65 @@
...
 from worker import conn, redis_url
+import time
+import json
+from flask_socketio import SocketIO
 
 q = Queue(connection=conn)
 
 
-app = Flask(__name__)
+app = Flask(__name__, static_url_path='')
```
Comment (abrichr) [inline app.py:18]: Why is this necessary? Can't we just leave it at `static`?

File: app.py [hunk 3]
```diff
@@ -2,27 +2,65 @@
...
+    '''
+    This function updates the meta dictionary of a given job and returns the
+    job's filename
+    '''
+    if job.meta.get('status') != 'Done':
+        try:
+            job.refresh()
+            print('Job refreshed successfully')
+        except:
+            pass
```
Comment (abrichr) [inline app.py]: Please specify which exception you are catching and print it to stdout.

File: app.py [hunk 4]
```diff
@@ -2,27 +2,65 @@
...
 app.register_blueprint(
     rq_dashboard.blueprint, url_prefix='/rq/'
 )
-
 jobs = []
 
+socket = SocketIO(app, cors_allowed_origins='*',  message_queue=os.getenv('REDIS_URL')) 
+
+@app.template_filter('job_refresh')
+def job_refresh(job):
```
Comment (abrichr) [inline app.py]: Please rename to `refresh_job`

File: app.py [hunk 5]
```diff
@@ -2,27 +2,65 @@
...
+    return job.filename
+
+@app.template_filter('jsonify_data')
+def jsonify_data(data):
+    '''
+    This job converts an object to a string containing that object
+    '''
+    try:
+        return json.dumps(data)
+    except:
```
Comment (abrichr) [inline app.py]: Please specify which exception(s) you are catching

File: app.py [hunk 6]
```diff
@@ -2,27 +2,65 @@
...
+        return json.dumps('Beginning process...')
+
+@app.route('/videos/<path:path>')
+def send_video(path):
+    return send_from_directory('videos', path)
+
+@app.route('/sample_video/<path:path>')
+def send_sample_video(path):
+    return send_from_directory('sample_video', path)
```
Comment (abrichr) [inline app.py]: I believe we only have one sample video. Can you please just send that file instead of a creating a whole directory and sending it?

File: app.py [hunk 7]
```diff
@@ -2,27 +2,65 @@
...
+
+@app.route('/spin/<path:path>')
+def send_spinner(path):
```
Comment (abrichr) [inline app.py]: Please do not add a new route for every third party dependency.  Just put the necesary files under the `static` path, e.g.:

```
- static
  - css
    - lib
      - spin.css
      - ...
    - ...
  - js
    - lib
      - spin.js
      - ...
    - ...
  - images
  - videos
```

File: app.py [hunk 8]
```diff
@@ -34,17 +72,23 @@ def upload():
     video_stream = video_file.read()
     with open(fname_video, 'wb') as f:
       f.write(video_stream)
-
+
```
Comment (abrichr) [inline app.py:78]: Please remove spurious whitespace.

See here for how to do this easily in Vim: https://vim.fandom.com/wiki/Remove_unwanted_spaces

File: app.py [hunk 9]
```diff
@@ -34,17 +72,23 @@ def upload():
...
         timeout=one_week
     )
+    
     job.filename = fname_video
+    job.tracks_filename = output_file 
     jobs.append(job)
+    
     logger.info(f'job: {job}')
-
+
```
Comment (abrichr) [inline app.py:93]: Remove spurious whitespace

File: docker-compose.yml [hunk 10]
```diff
@@ -1,17 +1,20 @@
-version: '2'
+version: '3.7'
```
Comment (abrichr) [inline docker-compose.yml:1]: What is the reason for this change?

File: docker-compose.yml [hunk 11]
```diff
@@ -1,17 +1,20 @@
...
     depends_on:
       - db
     volumes:
       - ./:/app
+    ports:
+      - "5000:5000"
   worker:
+    device_requests:
+      - capabilities:
+        - "gpu"
```
Comment (abrichr) [inline docker-compose.yml:17]: Can you please confirm that this will still work if the device doesn't have a gpu?

File: docker-compose.yml [hunk 12]
```diff
@@ -23,9 +26,8 @@ services:
   db:
     image: postgres:latest
     env_file: .env
-    ports:
-      - "5432:5432"
   redis:
     image: redis:alpine
-    ports:
-      - "6379:6379"
+    command: redis-server --save ""
```
Comment (abrichr) [inline docker-compose.yml:31]: What is the reason for this change?

File: docker-compose.yml [hunk 13]
```diff
@@ -23,9 +26,8 @@ services:
...
     env_file: .env
-    ports:
-      - "5432:5432"
   redis:
     image: redis:alpine
-    ports:
-      - "6379:6379"
+    command: redis-server --save ""  
+    volumes:
+      - "./redis-data:/data"
```
Comment (abrichr) [inline docker-compose.yml:33]: What is the reason for this change?

File: spin/spin.css [hunk 14]
```diff
@@ -0,0 +1,38 @@
+@keyframes spinner-line-fade-more {
```
Comment (abrichr) [inline spin/spin.css]: Please move to `static/css/lib`

File: spin/spin.js [hunk 15]
```diff
@@ -0,0 +1,187 @@
+var __assign = (this && this.__assign) || function () {
```
Comment (abrichr) [inline spin/spin.js]: Please move to `static/js/lib`

File: templates/index.html [hunk 16]
```diff
@@ -3,6 +3,22 @@
...
+    window.dataLayer = window.dataLayer || [];
+    function gtag(){dataLayer.push(arguments);}
+    gtag('js', new Date());
+    gtag('config', 'UA-166920629-3');
+  </script>
+
+  <script>
+    // temporary until Heroku bug gets sorted out
+    if (window.location.href.includes('heroku') && ! window.location.href.includes('-dev')) {
+      window.location.replace('http://ec2-18-234-105-116.compute-1.amazonaws.com:5000/')
```
Comment (abrichr) [inline templates/index.html]: Please update to GPU instance:

http://ec2-3-237-25-198.compute-1.amazonaws.com:5000/

File: templates/index.html [hunk 17]
```diff
@@ -80,30 +105,33 @@ <h1 class="ui header">
...
         {% for job in jobs %}
-          <li>
-            <ul>
-              <li>
-                filename: {{ job.filename }}
-              </li>
+            <ul style="list-style-type:none;">
               <li>
-                is_finished: {{ job.is_finished }}
+              {{ job | job_refresh }}
```
Comment (abrichr) [inline templates/index.html]: Is this necessary?

File: templates/index.html [hunk 18]
```diff
@@ -80,30 +105,33 @@ <h1 class="ui header">
...
-                result: {{ job.result }}
+                {% if job.meta.get('status') != 'Done' %}
+                  {% if job.meta.get('status') %}
+                    <div id="{{ job.id }}">{{ job.meta.get('status') }}</div>
+                  {% else %}
+                    <div id="{{ job.id }}">Beginning process...</div>
+                  {% endif %}
+                {% else %}
+                  <div id="{{ job.id }}">
+                    <video width="320" height="240" style="margin-left:auto;margin-right:auto;display:block" controls>
```
Comment (abrichr) [inline templates/index.html]: Please move styles to a <style> tag

File: templates/index.html [hunk 19]
```diff
@@ -145,5 +173,50 @@ <h3>Videos</h3>
...
-
+  <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
+  <script src="https://cdnjs.cloudflare.com/ajax/libs/socket.io/2.0.4/socket.io.js"></script>
+  <script type='module'>
+    import { Spinner } from './spin/spin.js'
+    var socket = io.connect('http://ec2-3-237-25-198.compute-1.amazonaws.com:5000/');
+    
+    {% for job in jobs %}
+      var spinner = new Spinner().spin(document.getElementById('{{ job.id }}'));
+    {% endfor %}
```
Comment (abrichr) [inline templates/index.html:174]: Remember, this will result in Javascript code that looks like this:

```
var spinner = ...
var spinner = ...
var spinner = ...
...
```
It doesn't make sense to assign this to a variable if you are just going to overwrite it immediately afterwards. At the very least, please remove `var spinner = `. I would do it like this:
```
let jobs = {% jobs|safe %}
for (let job of jobs) { ... }
```

File: templates/index.html [hunk 20]
```diff
@@ -145,5 +173,50 @@ <h3>Videos</h3>
...
+    var spinners = document.getElementsByClassName('spinner');
+    for(var i=0; i < spinners.length; i++) {
```
Comment (abrichr) [inline templates/index.html]: Please add a space after `for`

File: templates/index.html [hunk 21]
```diff
@@ -145,5 +173,50 @@ <h3>Videos</h3>
...
+     spinners[i].style.removeProperty('top');
```
Comment (abrichr) [inline templates/index.html]: What is this for?

File: templates/index.html [hunk 22]
```diff
@@ -145,5 +173,50 @@ <h3>Videos</h3>
...
+      console.log(job_status);
+      console.log(job_id);
+      if (typeof status === 'undefined') {
+        var innerHTML = 'Beginning process...';
+      } else {
+        var innerHTML =  job_status;
+      }
+      if (innerHTML === 'Done') {
+        document.getElementById(job_id).innerHTML = `
+        <video width="320" height="240" style="margin-left:auto;margin-right:auto;display:block" controls>
```
Comment (abrichr) [inline templates/index.html:200]: Please add styles to a <style> tag

File: templates/index.html [hunk 23]
```diff
@@ -145,5 +173,50 @@ <h3>Videos</h3>
...
+          <source src="/videos/${job_fname}" type="video/mp4">
+        </video>
+        `;
+      } else {
+        var target = document.getElementById(job_id);
+        target.innerHTML = innerHTML;
+        var spinner = new Spinner().spin(target);
+        var spinners = document.getElementsByClassName('spinner');                     
+        for(var i=0; i < spinners.length; i++) {                                       
+          spinners[i].style.removeProperty('top');
```
Comment (abrichr) [inline templates/index.html]: What is this for?

File: templates/index.html [hunk 24]
```diff
@@ -145,5 +173,50 @@ <h3>Videos</h3>
...
+        <video width="320" height="240" style="margin-left:auto;margin-right:auto;display:block" controls>
+          <source src="/videos/${job_fname}" type="video/mp4">
+        </video>
+        `;
+      } else {
+        var target = document.getElementById(job_id);
+        target.innerHTML = innerHTML;
+        var spinner = new Spinner().spin(target);
+        var spinners = document.getElementsByClassName('spinner');                     
+        for(var i=0; i < spinners.length; i++) {
```
Comment (abrichr) [inline templates/index.html]: Please add spaces after `for` and around `=`:

`for (var i = 0; ...)`

File: tracking.py [hunk 25]
```diff
@@ -797,7 +797,7 @@ def compute_matches(tracks, t, max_age,
   Returns:
       matches: A 1D np.ndarray with as many elements as boxes in current
       frame (cur_boxes). For each, there is an integer to index the previous
-      frame box that it matches to, or -1 if it doesnot match to any previous
+      frame box that it matches to, or -1 if it doesi not match to any previous
```
Comment (abrichr) [inline tracking.py:800]: Typo

File: tracking.py [hunk 26]
```diff
@@ -840,21 +840,21 @@ def compute_matches(tracks, t, max_age,
...
-def update_tracks(tracks, matches, t, all_boxes, all_keypoints):
-  """Updates the tracks for frame t given the matches, and creates new tracks when necessary.
+def update_tracks(tracks, matches, i, all_boxes, all_keypoints):
+  """Updates the tracks for frame i given the matches, and creates new tracks when necessary.
 
   Args:
   """
-  for i in range(len(matches)):
-    idx = matches[i]
+  print(matches)
```
Comment (abrichr) [inline tracking.py:848]: Please add a label to provide context to the reader, e.g.

```
printf('matches:', matches)
```

File: vision.py [hunk 27]
```diff
@@ -10,15 +11,72 @@
 from tqdm import tqdm
 from PIL import Image
 import tracking
+import time
```
Comment (abrichr) [inline vision.py:14]: Please add standard library imports to the top import block, and third party imports to the second import block.

First party imports (i.e. from this project) should go into a third block.

File: vision.py [hunk 28]
```diff
@@ -10,15 +11,72 @@
...
+    - str fpath_video: path to video file
+    - str output_file: path to output file (doesn't need to already exist)
+    '''
+    print('Cuda available {}'.format(torch.cuda.is_available()))
+    job = get_current_job()
+    job_id = str(job.id)
+    pd = 'progress display'
+
+    # Opening socket
+    socketio = SocketIO(message_queue=os.getenv('REDIS_URL'),
```
Comment (abrichr) [inline vision.py:46]: If the line needs to be broken up, please format like:

```
socketio = SocketIO(
    message_queue=os.getenv('REDIS_URL'), 
    cors_allowed_origins='*'
)

File: vision.py [hunk 29]
```diff
@@ -10,15 +11,72 @@
...
+    Given the path to a video, creates a tracking video and writes it to the
+    path output_file
+    Parameters:
+    - str fpath_video: path to video file
+    - str output_file: path to output file (doesn't need to already exist)
+    '''
+    print('Cuda available {}'.format(torch.cuda.is_available()))
+    job = get_current_job()
+    job_id = str(job.id)
+    pd = 'progress display'
```
Comment (abrichr) [inline vision.py]: Please name this what it is, not what it contains, e.g. `message_type`

File: vision.py [hunk 30]
```diff
@@ -10,15 +11,72 @@
...
+    job.meta['status'] = status
+    job.save()
+    print('Sent "Getting getting predictions from images" through socket')  
     predictions = _get_predictions_from_images(images)
+    
+    # Getting tracks from predictions
+    status = 'Getting tracks from predictions'
+    socketio.emit(pd, {'status': status, 'id': job_id}, json=True)        
+    job.meta['status'] = status        
+    job.save()
```
Comment (abrichr) [inline vision.py]: DRY: Don't Repeat Yourself

You've used this pattern several times:
```
    status = 'Getting tracks from predictions'
    socketio.emit(pd, {'status': status, 'id': job_id}, json=True)        
    job.meta['status'] = status        
    job.save()
```

Please move this to a function

File: vision.py [hunk 31]
```diff
@@ -10,15 +11,72 @@
...
+    status = 'Making video from tracks'
+    socketio.emit(pd, {'status': status, 'id': job_id}, json=True)        
+    job.meta['status'] = status        
+    job.save()
+    print('Sent "Making video from tracks" through socket')  
+    fpath_tracking_video = _get_video_from_tracks(tracks, images, output_file)
+    
+    # Done
+    status = 'Done'
+    socketio.emit(pd, {'status': status, 'id': job_id,
```
Comment (abrichr) [inline vision.py]: Formatting as above

File: vision.py [hunk 32]
```diff
@@ -104,28 +162,48 @@ def _setup_cfg(config, opts, conf_thresh):
...
 
-def _get_video_from_tracks(tracks, images):
+def _get_video_from_tracks(tracks, images, output_file):
     ''' Save a video showing tracks to disk and return the path '''
-    
-    # fourcc = cv2.VideoWriter_fourcc(*'MP4V')
     output_size = images[0].shape
-    output_file = 'tracks.mp4'
-    # out = cv2.VideoWriter(output_file, fourcc, 25, output_size[:-1])
+    kelly_colors_rgb = [(255, 179, 0), (128, 62, 117), (255, 104, 0), (166, 189, 215),
```
Comment (abrichr) [inline vision.py:165]: Please add a comment containing an explanation and/or link to a source

File: vision.py [hunk 33]
```diff
@@ -104,28 +162,48 @@ def _setup_cfg(config, opts, conf_thresh):
...
-        frame.save('image_folder/frame{}.jpg'.format(i)) 
-        # out.write(frame.astype(np.uint8))
-
-    # out.release()
+        image_file = '{}/frame{}.jpg'.format(image_folder, i)
+        image_files.append(image_file)
+        frame.save(image_file) 
+    
+    
+    clip = moviepy.video.io.ImageSequenceClip.ImageSequenceClip(image_files, fps=15)
```
Comment (abrichr) [inline vision.py:196]: Why 15 fps?

File: web.requirements.txt [hunk 34]
```diff
@@ -3,3 +3,6 @@ flask==1.1.2
 requests==2.24.0
 rq==1.4.3
 rq_dashboard==0.6.1
+boto3==1.14.50
+flask-socketio
+eventlet
```
Comment (abrichr) [inline web.requirements.txt]: Please add versions

File: web.sh [hunk 35]
```diff
@@ -1 +1 @@
-gunicorn -b 0.0.0.0:$PORT app:app
+gunicorn -b 0.0.0.0:$PORT --worker-class eventlet -w 1 app:app --timeout 500
```
Comment (abrichr) [inline web.sh:4]: Please add a comment above containing an explanation and link to documentation regarding `worker-class`

File: worker.Dockerfile [hunk 36]
```diff
@@ -1,10 +1,29 @@
-FROM jjanzic/docker-python3-opencv
+# FROM jjanzic/docker-python3-opencv
```
Comment (abrichr) [inline worker.Dockerfile]: Please remove if not in use

File: worker.Dockerfile [hunk 37]
```diff
@@ -1,10 +1,29 @@
...
 
 WORKDIR /app
 COPY . /app
-RUN pip install --upgrade cython && pip install -r worker.requirements.txt && python -m pip install 'git+https://github.com/facebookresearch/detectron2.git'
-CMD python worker.py
+
+RUN pip3 install --upgrade pip
+RUN pip3 install PyYAML
+
+RUN pip3 install numpy
```
Comment (abrichr) [inline worker.Dockerfile]: Why are these happening here and not in `requirements.txt`?

File: worker.requirements.txt [hunk 38]
```diff
@@ -7,4 +7,8 @@ rq_dashboard==0.6.1
 scipy
 torch==1.6
 torchvision==0.7
+numpy==1.19.1
 tqdm
+moviepy
+boto3==1.14.50
+flask-socketio
```
Comment (abrichr) [inline worker.requirements.txt]: Please add version numbers

**Review by abrichr — CHANGES_REQUESTED** [review]
Thanks for putting this together Andrew! Looks good so far.

In addition to my inline comments, please remove all binary files, i.e. images and videos.

File: app.py (hunk 3 above)
Comment (DrJessop) [reply app.py]: There was a way to bypass this with conditionals instead of exceptions (line 35 of app.py)

File: worker.Dockerfile (hunk 37 above)
Comment (DrJessop) [reply worker.Dockerfile]: The file now installs numpy, cython, and then requirements.txt in that order because cython requires numpy as a dependency, and cython bbox in the requirements file requires cython as a dependency

File: templates/index.html (hunk 19 above)
Comment (DrJessop) [reply templates/index.html:174]: Please see line 185 of templates/index.html

File: .env (hunk 1 above)
Comment (DrJessop) [reply .env:3]: This was in the original master branch .env file, RQ was not able to connect without it before.

File: app.py (hunk 7 above)
Comment (DrJessop) [reply app.py]: This is done

File: vision.py (hunk 33 above)
Comment (DrJessop) [reply vision.py:196]: The sample video is exactly 30 frames long, and I didn't the tracking video to zoom by, so I thought that a 2 second video is appropriate
